server/app/api/v1/charity/raffle/controller.js

```diff
@@ -0,0 +1,219 @@
+    .catch(err => next(err));
+};
+
+/**
+ * Returns req.raffle
+ */
+const get = (req, res) => res.json(req.raffle);
+
+/**
+ * Creates new raffle
+ */
+const create = (req, res, next) => {
+  return new Raffle({
+    charity: req.charity._id,
+    prize: {
+      name: req.body.prizeName,
+      thumbnailURL: req.body.prizeThumbnailURL
+    }
+  }).save()
+  .then((savedRaffle) => {
+    req.charity.raffles.push(savedRaffle);
+    req.charity.save();
+    savedRaffle = savedRaffle.toObject();
+    return res.json(savedRaffle);
+  })
+  .catch(err => next(err));
+};
+
+/**
+ * Purchase raffle
+ * if existing user use query userId={userId}
+ * if new user user email, firstName, lastName in req.body
+ */
+
+const purchase = (req, res, next) => {
+  const hasUserId = req.query.hasOwnProperty('userId');
+  if (req.raffle.isActive) {
+    const findOrCreateUser = hasUserId ? findByUserId : createUser;
+
+    findOrCreateUser()
+      .then(addTickets)
+        .all()
+      .then(([raffle, user]) => {
+        const { email, firstName, lastName } = user;
+        const { count } = raffle.purchases.filter(
+          purchase => purchase.buyer === user._id
+        )[0];
+        return res.json({
+          count,
+          user: {
+            email,
+            firstName,
+            lastName
+          }
+        });
+      })
+      .catch(err => next(err));
+  } else {
+    next(
+      new NotPermittedError(
+        'Tickets cannot be purchased after raffle has ended.'
+      )
+    );
+  }
+
+  function findByUserId() {
+    const userId = req.query.userId;
+
+    if (!userId.match(VALID_OBJECT_ID)) {
+      return next(
+        new ValidationError(
+          'charityId must be valid ObjectId',
+          VALID_OBJECT_ID,
+          'charityId'
+        )
+      );
+    }
+    return User.findById(userId)
+      .then((user) => {
+        if (!user) {
+          throw new NotFoundError('User not found');
+        } else {
+          return user;
+        }
+      });
+  }
+
+  function createUser() {
+    return new User({
+      firstName: req.body.firstName,
+      lastName: req.body.lastName,
+      email: req.body.email
+    }).save();
+  }
+
+  function addTickets(user) {
+    // checks for if user already purchased raffles
+    const purchaseIndex = findIndex(req.raffle.purchases, ['buyer', user._id]);
+    const userRaffles = find(user.raffles, req.raffle._id);
+
+    if (purchaseIndex > -1) {
+      // remove duplicate purchase from purchases array
+      // this is a workaround to an Mongoose error I was receiving
+      // See: https://github.com/Automattic/mongoose/issues/3816
+      const dupPurchase = req.raffle.purchases.splice(purchaseIndex, 1)[0];
+      req.raffle.purchases.push({
+        count: dupPurchase.count + Number(req.body.count),
+        buyer: user._id
+      });
+    } else {
+      req.raffle.purchases.push({
+        count: Number(req.body.count),
+        buyer: user._id
+      });
+    }
```

Comment:
More verbose, but clearer way to write this:

```javascript
const newPurchase = {
  count: Number(req.body.count),
  buyer: user._id
};

const purchaseIndex = findIndex(req.raffle.purchases, ['buyer', user._id]);
if (purchaseIndex > -1) {
  const dupPurchase = req.raffle.purchases.splice(purchaseIndex, 1)[0];
  newPurchase.count += dupPurchase.count
}

req.raffle.purchases.push(newPurchase);
```

It's DRYer and a bit clearer. In this case, the pushed object only has 2 properties so both ways are good. If it has 3+, use a new variable.